// @slidev/parser@0.43.7/dist/core.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});
















var _chunkHBQPSWBSjs = require('./chunk-HBQPSWBS.js');

















exports.detectFeatures = _chunkHBQPSWBSjs.detectFeatures; exports.filterDisabled = _chunkHBQPSWBSjs.filterDisabled; exports.getDefaultConfig = _chunkHBQPSWBSjs.getDefaultConfig; exports.mergeFeatureFlags = _chunkHBQPSWBSjs.mergeFeatureFlags; exports.parse = _chunkHBQPSWBSjs.parse; exports.parseAspectRatio = _chunkHBQPSWBSjs.parseAspectRatio; exports.parseRangeString = _chunkHBQPSWBSjs.parseRangeString; exports.parseSlide = _chunkHBQPSWBSjs.parseSlide; exports.prettify = _chunkHBQPSWBSjs.prettify; exports.prettifySlide = _chunkHBQPSWBSjs.prettifySlide; exports.resolveConfig = _chunkHBQPSWBSjs.resolveConfig; exports.resolveFonts = _chunkHBQPSWBSjs.resolveFonts; exports.scanMonacoModules = _chunkHBQPSWBSjs.scanMonacoModules; exports.stringify = _chunkHBQPSWBSjs.stringify; exports.stringifySlide = _chunkHBQPSWBSjs.stringifySlide; exports.verifyConfig = _chunkHBQPSWBSjs.verifyConfig;
